056-zk-kraft-migration.md

```diff
@@ -0,0 +1,258 @@
+
+From this state it's still possible to not move forward with the migration.
+The user can delete the `KafkaNodePool` custom resource containing the definition of the KRaft controllers and remove the `strimzi.io/kraft-migration` annotation from the `Kafka` custom resource.
+The FSM moves back to the `NoMigration` state (or null) actually exiting from the migration process and the operator removes the `strimzi.io/kraft-migration-state` annotation from the `Kafka` custom resource.
+There is nothing to do with the brokers which are still configured to work with the ZooKeeper ensemble only.
+
+#### Phase 2: Enabling the migration on the brokers
+
+In the `WaitBrokersInMigration` state, the operator expects the user to ask for running the actual metadata migration.
+In order to start the actual migration, the brokers need to be connected to the KRaft controllers and have the ZooKeeper migration enabled.
+The user annotates the `Kafka` custom resource running the current ZooKeeper based brokers with `strimzi.io/kraft-migration: migration`.
+This annotation drives the operator to update the brokers' configuration with the ZooKeeper migration flag enabled and with the new KRaft controllers quorum connection information as well.
+
+```shell
+# KRaft controller quorum configuration
+controller.quorum.voters=<voters>
+controller.listener.names=CONTROLLER
+
+# The controller.listener.name should also be added to listener.security.property.map
+listener.security.protocol.map=PLAINTEXT:PLAINTEXT,CONTROLLER:PLAINTEXT
+
+# Enable the migration
+zookeeper.metadata.migration.enable=true
+```
+
+Finally, the operator rolls the brokers.
+After the rolling, the internal Apache Kafka migration component starts the actual metadata migration and the operator has to wait for it to be completed to move to the next phase.
+Checking the migration progress can be done using some of the exposed migration metrics.
+The FSM moves to `Migration` state first, while migration is ongoing, and finally to the `DualWrite` state.
+In both states, the user can rollback the process by applying the `strimzi.io/kraft-migration: pre-migration` on the `Kafka` custom resource.
+In this case, the operator configures the brokers with no ZooKeeper migration flag, and no KRaft controllers quorum connection information.
+Finally, the operator rolls the brokers and wait for them to be ready again.
+The FSM moves back to the `WaitBrokersInMigration`, so actually at the beginning of the Phase 2.
+
+If the user doesn't rollback the migration, next step is to move brokers and controllers out of migration and using KRaft only in Phase 3/4.
+
+#### Phase 3/4: Migrating brokers and controllers to full KRaft
+
+In the `DualWrite` state, the operator expects the user desire to finalize the migration, having a full KRaft based cluster and ZooKeeper out of the picture.
+The user applies the `strimzi.io/kraft-migration: post-migration` annotation to the `Kafka` custom resource.
```

Comment:
Again, how does the user find out how to do it?

Reply:
After latest change, there is the `Kafka.status.migrationState` the user has to check.